src/core/currency/wallet/currency-wallet-cleaners.ts

```diff
@@ -246,7 +254,37 @@ export const asEdgeTxActionStake = asObject({
   stakeAssets: asArray(asEdgeAssetAmount)
 })
 
-export const asEdgeTxAction = asEither(asEdgeTxActionSwap, asEdgeTxActionStake)
+export const asEdgeTxActionFiatType: Cleaner<EdgeTxActionFiatType> = asValue(
+  'buy',
+  'sell',
+  'sellNetworkFee'
```

Comment:
OK, so with this design, we need to save two copies of the action, duplicating the orderId and all other fields. One copy goes with the parent currency as a fee, and one goes with the token. The only difference between the two copies is the `type` field.

You already have a `tokenId` inside the action itself, so there is no need for this duplication. The GUI can easily see which side is the fee and which side is the sell by using `tx.tokenId === savedAction.cryptoAsset.tokenId`.

So, I really think the whole concept of saving the action per-currency is a mistake. There should be one action per transaction, saved at the root level.

When we first discussed this, I trusted your assertion that the actions needed to be saved per-token. Now that I'm seeing what the action looks like, I think that's a deeply flawed assumption.

Reply:
In the case of a fiat transaction this might be simple, but consider a DEX swap on Ripple. You have an action to create an order which pays XRP, the outgoing send which is considered a 'swap' which could pay USD.bs, then the receive which is the fulfillment that receives USD.gh. Do you really want the GUI to have to figure that all out.

As well the plugins report transactions per asset as they send individual EdgeTransactions. Now while they might not use `savedAction` but instead `action`, they would still be putting the action info per asset which we don't want to conflict.